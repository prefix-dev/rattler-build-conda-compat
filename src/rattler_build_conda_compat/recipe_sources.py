from __future__ import annotations

import sys
import typing
from typing import Any, List, TypedDict, Union

from .conditional_list import ConditionalList, visit_conditional_list

if sys.version_info < (3, 11):
    from typing_extensions import NotRequired
else:
    from typing import NotRequired

if typing.TYPE_CHECKING:
    from collections.abc import Iterator, Mapping

OptionalUrlList = Union[str, List[str], None]


class Source(TypedDict):
    url: NotRequired[str]


def get_all_url_sources(recipe: Mapping[Any, Any]) -> Iterator[str]:
    """
    Get all url sources from the recipe. This can be from a list of sources,
    a single source, or conditional and its branches.

    Arguments
    ---------
    * `recipe` - The recipe to inspect. This should be a yaml object.

    Returns
    -------
    A list of sources.
    """

    sources = recipe.get("source", None)
    sources = typing.cast(ConditionalList[Source], sources)

    # Try getting all url top-level sources
    if sources is not None:
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url

    outputs = recipe.get("outputs", None)
    if outputs is None:
        return

    outputs = visit_conditional_list(outputs, None)
    for output in outputs:
        sources = output.get("source", None)
        sources = typing.cast(ConditionalList[Source], sources)
        if sources is None:
            continue
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url
