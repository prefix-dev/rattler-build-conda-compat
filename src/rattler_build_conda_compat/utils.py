import fnmatch
import itertools
from logging import getLogger
import os
from pathlib import Path
from typing import Any, Dict, Iterable


VALID_METAS = ("recipe.yaml",)


def islist(arg, uniform=False, include_dict=True):
    """
    Check whether `arg` is a `list`. Optionally determine whether the list elements
    are all uniform.

    When checking for generic uniformity (`uniform=True`) we check to see if all
    elements are of the first element's type (`type(arg[0]) == type(arg[1])`). For
    any other kinds of uniformity checks are desired provide a uniformity function:

    .. code-block:: pycon
        # uniformity function checking if elements are str and not empty
        >>> truthy_str = lambda e: isinstance(e, str) and e
        >>> islist(["foo", "bar"], uniform=truthy_str)
        True
        >>> islist(["", "bar"], uniform=truthy_str)
        False
        >>> islist([0, "bar"], uniform=truthy_str)
        False

    .. note::
        Testing for uniformity will consume generators.

    :param arg: Object to ensure is a `list`
    :type arg: any
    :param uniform: Whether to check for uniform or uniformity function
    :type uniform: bool, function, optional
    :param include_dict: Whether to treat `dict` as a `list`
    :type include_dict: bool, optional
    :return: Whether `arg` is a `list`
    :rtype: bool
    """
    if isinstance(arg, str) or not isinstance(arg, Iterable):
        # str and non-iterables are not lists
        return False
    elif not include_dict and isinstance(arg, dict):
        # do not treat dict as a list
        return False
    elif not uniform:
        # short circuit for non-uniformity
        return True

    # NOTE: not checking for Falsy arg since arg may be a generator
    # WARNING: if uniform != False and arg is a generator then arg will be consumed

    if uniform is True:
        arg = iter(arg)
        try:
            etype = type(next(arg))
        except StopIteration:
            # StopIteration: list is empty, an empty list is still uniform
            return True
        # check for explicit type match, do not allow the ambiguity of isinstance
        uniform = lambda e: type(e) == etype  # noqa: E731

    try:
        return all(uniform(e) for e in arg)
    except (ValueError, TypeError):
        # ValueError, TypeError: uniform function failed
        return False


def ensure_list(arg, include_dict=True):
    """
    Ensure the object is a list. If not return it in a list.

    :param arg: Object to ensure is a list
    :type arg: any
    :param include_dict: Whether to treat `dict` as a `list`
    :type include_dict: bool, optional
    :return: `arg` as a `list`
    :rtype: list
    """
    if arg is None:
        return []
    elif islist(arg, include_dict=include_dict):
        return list(arg)
    else:
        return [arg]


def rec_glob(path, patterns, ignores=None):
    """
    Recursively searches path for filename patterns.

    :param path: path within to search for files
    :param patterns: list of filename patterns to search for
    :param ignore: list of directory patterns to ignore in search
    :return: list of paths in path satisfying patterns/ignore
    """
    patterns = ensure_list(patterns)
    ignores = ensure_list(ignores)

    for path, dirs, files in os.walk(path):
        # remove directories to ignore
        for ignore in ignores:
            for d in fnmatch.filter(dirs, ignore):
                dirs.remove(d)

        # return filepaths that match a pattern
        for pattern in patterns:
            for f in fnmatch.filter(files, pattern):
                yield os.path.join(path, f)


def find_recipe(path):
    """
    vendored from conda_build.utils to persist same API flow

    recurse through a folder, locating valid meta files (see VALID_METAS).  Raises error if more than one is found.

    Returns full path to meta file to be built.

    If we have a base level meta file and other supplemental (nested) ones, use the base level.
    """
    # if initial path is absolute then any path we find (via rec_glob)
    # will also be absolute
    if not os.path.isabs(path):
        path = os.path.normpath(os.path.join(os.getcwd(), path))

    if os.path.isfile(path):
        if os.path.basename(path) in VALID_METAS:
            return path
        raise OSError("{} is not a valid meta file ({})".format(path, ", ".join(VALID_METAS)))

    results = list(rec_glob(path, VALID_METAS, ignores=(".AppleDouble",)))

    if not results:
        raise OSError("No meta files ({}) found in {}".format(", ".join(VALID_METAS), path))

    if len(results) == 1:
        return results[0]

    # got multiple valid meta files
    # check if a meta file is defined on the base level in which case use that one

    metas = [m for m in VALID_METAS if os.path.isfile(os.path.join(path, m))]
    if len(metas) == 1:
        getLogger(__name__).warn(
            "Multiple meta files found. "
            f"The {metas[0]} file in the base directory ({path}) "
            "will be used."
        )
        return os.path.join(path, metas[0])

    raise OSError("More than one meta files ({}) found in {}".format(", ".join(VALID_METAS), path))


def has_recipe(recipe_dir: Path) -> bool:
    """

    verify if recipe_dir contains recipe.yaml

    it is used to detect whetever old or recipe should be used


    """
    try:
        recipe_path = find_recipe(recipe_dir)
        if recipe_path:
            return True
        return False
    except OSError:
        return False


def remove_empty_keys(some_dict: Dict[str, Any]) -> Dict[str, Any]:
    filtered_dict = {}
    for key, value in some_dict.items():
        if isinstance(value, list) and len(value) == 0:
            continue
        filtered_dict[key] = value

    return filtered_dict


def flatten_lists(some_dict: Dict[str, Any]) -> Dict[str, Any]:
    result_dict: Dict[str, Any] = {}
    for key, value in some_dict.items():
        if isinstance(value, dict):
            result_dict[key] = flatten_lists(value)
        elif isinstance(value, list) and value and isinstance(value[0], list):
            result_dict[key] = list(itertools.chain(*value))
        else:
            result_dict[key] = value

    return result_dict
