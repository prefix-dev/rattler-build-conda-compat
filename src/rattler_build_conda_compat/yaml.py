import io
from typing import Any

from ruamel.yaml import YAML
from ruamel.yaml.comments import CommentedMap, CommentedSeq


# Custom constructor for loading floats as strings
def float_as_string_constructor(loader, node) -> str:  # noqa: ANN001
    return loader.construct_scalar(node)


def convert_to_plain_types(data: Any) -> Any:  # noqa: ANN401
    if isinstance(data, CommentedMap):
        return {convert_to_plain_types(k): convert_to_plain_types(v) for k, v in data.items()}
    if isinstance(data, (CommentedSeq, list)):
        return [convert_to_plain_types(item) for item in data]
    if isinstance(data, dict):
        return {convert_to_plain_types(k): convert_to_plain_types(v) for k, v in data.items()}
    return data


def _yaml_object() -> YAML:
    yaml = YAML(typ="rt")
    yaml.Constructor.add_constructor("tag:yaml.org,2002:float", float_as_string_constructor)
    yaml.allow_duplicate_keys = False
    yaml.preserve_quotes = True
    yaml.width = 320
    yaml.indent(mapping=2, sequence=4, offset=2)
    return yaml


def _dump_yaml_to_string(data: Any) -> str:  # noqa: ANN401
    yaml = _yaml_object()
    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()
