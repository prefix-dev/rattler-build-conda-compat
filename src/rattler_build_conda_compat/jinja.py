from __future__ import annotations

from typing import Any, TypedDict

import jinja2
import yaml
from jinja2 import DebugUndefined

from rattler_build_conda_compat.loader import load_yaml


class RecipeWithContext(TypedDict, total=False):
    context: dict[str, str]


class _MissingUndefined(DebugUndefined):
    def __str__(self) -> str:
        """
        By default, `DebugUndefined` return values in the form `{{ value }}`.
        `rattler-build` has a different syntax, so we need to override this method,
        and return the value in the form `${{ value }}`.
        """
        return f"${super().__str__()}"


def jinja_env() -> jinja2.Environment:
    """
    Create a `rattler-build` specific Jinja2 environment with modified syntax.
    """
    return jinja2.Environment(
        variable_start_string="${{",
        variable_end_string="}}",
        trim_blocks=True,
        lstrip_blocks=True,
        autoescape=True,
        undefined=_MissingUndefined,
    )


def load_recipe_context(context: dict[str, str], jinja_env: jinja2.Environment) -> dict[str, str]:
    """
    Load all string values from the context dictionary as Jinja2 templates.
    """
    # Process each key-value pair in the dictionary
    for key, value in context.items():
        # If the value is a string, render it as a template
        if isinstance(value, str):
            template = jinja_env.from_string(value)
            rendered_value = template.render(context)
            context[key] = rendered_value

    return context


def render_recipe_with_context(recipe_content: RecipeWithContext) -> dict[str, Any]:
    """
    Render the recipe using known values from context section.
    Unknown values are not evaluated and are kept as it is.

    Examples:
    ---
    ```python
    >>> from pathlib import Path
    >>> from rattler_build_conda_compat.loader import load_yaml
    >>> recipe_content = load_yaml((Path().resolve() / "tests" / "data" / "eval_recipe_using_context.yaml").read_text())
    >>> evaluated_context = render_recipe_with_context(recipe_content)
    >>> assert "my_value-${{ not_present_value }}" == evaluated_context["build"]["string"]
    >>>
    ```
    """
    env = jinja_env()
    context = recipe_content.get("context", {})
    # load all context templates
    context_templates = load_recipe_context(context, env)

    # render the rest of the document with the values from the context
    # and keep undefined expressions _as is_.
    template = env.from_string(yaml.dump(recipe_content))
    rendered_content = template.render(context_templates)
    return load_yaml(rendered_content)
