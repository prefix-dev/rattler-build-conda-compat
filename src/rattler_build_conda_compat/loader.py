from __future__ import annotations

from contextlib import contextmanager
from typing import TYPE_CHECKING, Any

import yaml

from rattler_build_conda_compat.conditional_list import visit_conditional_list
from rattler_build_conda_compat.utils import flatten_lists, remove_empty_keys

if TYPE_CHECKING:
    from collections.abc import Iterator
    from os import PathLike

SELECTOR_OPERATORS = ("and", "or", "not")


class RecipeLoader(yaml.BaseLoader):
    _namespace: dict[str, Any] | None = None
    _allow_missing_selector: bool = False

    @classmethod
    @contextmanager
    def with_namespace(
        cls: type[RecipeLoader],
        namespace: dict[str, Any] | None,
        *,
        allow_missing_selector: bool = False,
    ) -> Iterator[None]:
        try:
            cls._namespace = namespace
            cls._allow_missing_selector = allow_missing_selector
            yield
        finally:
            del cls._namespace

    def construct_sequence(  # noqa: C901, PLR0912
        self,
        node: yaml.ScalarNode | yaml.SequenceNode | yaml.MappingNode,
        deep: bool = False,  # noqa: FBT002, FBT001,
    ) -> list[yaml.ScalarNode]:
        """deep is True when creating an object/mapping recursively,
        in that case want the underlying elements available during construction
        """
        # find if then else selectors
        for sequence_idx, child_node in enumerate(node.value[:]):
            # if then is only present in MappingNode

            if isinstance(child_node, yaml.MappingNode):
                # iterate to find if there is IF first

                the_evaluated_one = None
                for idx, (key_node, value_node) in enumerate(child_node.value):
                    if key_node.value == "if":
                        # we catch the first one, let's try to find next pair of (then | else)
                        then_node_key, then_node_value = child_node.value[idx + 1]

                        if then_node_key.value != "then":
                            msg = "cannot have if without then, please reformat your variant file"
                            raise ValueError(msg)

                        try:
                            _, else_node_value = child_node.value[idx + 2]
                        except IndexError:
                            _, else_node_value = None, None

                        to_be_eval = f"{value_node.value}"

                        if self._allow_missing_selector:
                            split_selectors = [
                                selector
                                for selector in to_be_eval.split()
                                if selector not in SELECTOR_OPERATORS
                            ]
                            for selector in split_selectors:
                                if self._namespace and selector not in self._namespace:
                                    cleaned_selector = selector.strip("(").rstrip(")")
                                    self._namespace[cleaned_selector] = True

                        evaled = eval(to_be_eval, self._namespace)  # noqa: S307
                        if evaled:
                            the_evaluated_one = then_node_value
                        elif else_node_value:
                            the_evaluated_one = else_node_value

                        if the_evaluated_one:
                            node.value.remove(child_node)
                            node.value.insert(sequence_idx, the_evaluated_one)
                        else:
                            # neither the evaluation or else node is present, so we remove this if
                            node.value.remove(child_node)

        if not isinstance(node, yaml.SequenceNode):
            raise TypeError(
                None,
                None,
                f"expected a sequence node, but found {node.id!s}",
                node.start_mark,
            )

        return [self.construct_object(child, deep=deep) for child in node.value]


def load_yaml(content: str | bytes) -> Any:  # noqa: ANN401
    return yaml.load(content, Loader=yaml.BaseLoader)  # noqa: S506


def parse_recipe_config_file(
    path: PathLike[str], namespace: dict[str, Any] | None, *, allow_missing_selector: bool = False
) -> dict[str, Any]:
    with open(path) as f, RecipeLoader.with_namespace(
        namespace, allow_missing_selector=allow_missing_selector
    ):
        content = yaml.load(f, Loader=RecipeLoader)  # noqa: S506
    return flatten_lists(remove_empty_keys(content))


def load_all_requirements(content: dict[str, Any]) -> dict[str, Any]:
    requirements_section = dict(content.get("requirements", {}))
    if not requirements_section:
        return {}

    for section in requirements_section:
        section_reqs = requirements_section[section]
        if not section_reqs:
            continue

        requirements_section[section] = list(visit_conditional_list(section_reqs))

    return requirements_section
