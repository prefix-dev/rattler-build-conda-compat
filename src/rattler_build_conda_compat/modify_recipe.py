from __future__ import annotations

import copy
import hashlib
import io
import logging
import re
from typing import TYPE_CHECKING, Any, Literal

import requests
from ruamel.yaml import YAML

from rattler_build_conda_compat.jinja.jinja import jinja_env, load_recipe_context
from rattler_build_conda_compat.recipe_sources import Source, get_all_sources

if TYPE_CHECKING:
    from pathlib import Path

logger = logging.getLogger(__name__)

yaml = YAML()
yaml.preserve_quotes = True
yaml.width = 4096
yaml.indent(mapping=2, sequence=4, offset=2)


def _update_build_number_in_context(recipe: dict[str, Any], new_build_number: int) -> bool:
    for key in recipe.get("context", {}):
        if key.startswith("build_") or key == "build":
            recipe["context"][key] = new_build_number
            return True
    return False


def _update_build_number_in_recipe(recipe: dict[str, Any], new_build_number: int) -> bool:
    is_modified = False
    if "build" in recipe and "number" in recipe["build"]:
        recipe["build"]["number"] = new_build_number
        is_modified = True

    if "outputs" in recipe:
        for output in recipe["outputs"]:
            if "build" in output and "number" in output["build"]:
                output["build"]["number"] = new_build_number
                is_modified = True

    return is_modified


def update_build_number(file: Path, new_build_number: int = 0) -> str:
    """
    Update the build number in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_build_number` - The new build number to use. (default: 0)

    Returns:
    --------
    The updated recipe as a string.
    """
    with file.open("r") as f:
        data = yaml.load(f)
    build_number_modified = _update_build_number_in_context(data, new_build_number)
    if not build_number_modified:
        _update_build_number_in_recipe(data, new_build_number)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()


class CouldNotUpdateVersionError(Exception):
    NO_CONTEXT = "Could not find context in recipe"
    NO_VERSION = "Could not find version in recipe context"

    def __init__(self, message: str = "Could not update version") -> None:
        self.message = message
        super().__init__(self.message)


class Hash:
    def __init__(self, hash_type: Literal["md5", "sha256"], hash_value: str) -> None:
        self.hash_type = hash_type
        self.hash_value = hash_value

    def __str__(self) -> str:
        return f"{self.hash_type}: {self.hash_value}"


def _has_jinja_version(url: str) -> bool:
    """Check if the URL has a jinja `${{ version }}` in it."""
    pattern = r"\${{\s*version"
    return re.search(pattern, url) is not None


def update_hash(source: Source, url: str, hash_: Hash | None) -> None:
    """
    Update the sha256 hash in the source dictionary.

    Arguments:
    ----------
    * `source` - The source dictionary to update.
    * `url` - The URL to download and hash (if no hash is provided).
    * `hash_` - The hash to use. If not provided, the file will be downloaded and `sha256` hashed.
    """
    hash_type = hash_.hash_type if hash_ is not None else "sha256"
    # delete all old hashes that we are not updating
    all_hash_types = {"md5", "sha256"}
    for key in all_hash_types - {hash_type}:
        if key in source:
            del source[key]

    if hash_ is not None:
        source[hash_.hash_type] = hash_.hash_value
    else:
        # download and hash the file
        hasher = hashlib.sha256()
        print(f"Retrieving and hashing {url}")
        with requests.get(url, stream=True, timeout=100) as r:
            for chunk in r.iter_content(chunk_size=4096):
                hasher.update(chunk)
        source["sha256"] = hasher.hexdigest()


def update_version(file: Path, new_version: str, hash_: Hash | None) -> str:
    """
    Update the version in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_version` - The new version to use.
    * `hash_type` - The hash type to use. If not provided, the file will be downloaded and `sha256` hashed.

    Returns:
    --------
    The updated recipe as a string.
    """

    with file.open("r") as f:
        data = yaml.load(f)

    if "context" not in data:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_CONTEXT)
    if "version" not in data["context"]:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_VERSION)

    data["context"]["version"] = new_version

    # set up the jinja context
    env = jinja_env()
    context = copy.deepcopy(data.get("context", {}))
    context_variables = load_recipe_context(context, env)
    # for r-recipes we add the default `cran_mirror` variable
    context_variables["cran_mirror"] = "https://cran.r-project.org"

    for source in get_all_sources(data):
        # render the whole URL and find the hash
        if "url" not in source:
            continue

        url = source["url"]
        if isinstance(url, list):
            url = url[0]

        if not _has_jinja_version(url):
            continue

        template = env.from_string(url)
        rendered_url = template.render(context_variables)

        update_hash(source, rendered_url, hash_)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()
