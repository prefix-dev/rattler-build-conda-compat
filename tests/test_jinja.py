from pathlib import Path

from rattler_build_conda_compat.jinja.filters import _version_to_build_string
from rattler_build_conda_compat.jinja.jinja import render_recipe_with_context
from rattler_build_conda_compat.jinja.utils import _MissingUndefined
from rattler_build_conda_compat.loader import load_yaml
from rattler_build_conda_compat.yaml import _dump_yaml_to_string

test_data = Path(__file__).parent / "data"


def test_render_recipe_with_context(snapshot) -> None:
    recipe = Path("tests/data/mamba_recipe.yaml")
    recipe_yaml = load_yaml(recipe.read_text())

    rendered = render_recipe_with_context(recipe_yaml)
    into_yaml = _dump_yaml_to_string(rendered)

    assert into_yaml == snapshot


def test_version_to_build_string() -> None:
    assert _version_to_build_string("1.2.3") == "12"
    assert _version_to_build_string("1.2") == "12"
    assert _version_to_build_string("nothing") == "nothing"
    some_undefined = _MissingUndefined(name="python")
    assert _version_to_build_string(some_undefined) == "python_version_to_build_string"


def test_context_rendering(snapshot) -> None:
    recipe = test_data / "context.yaml"

    recipe_yaml = load_yaml(recipe.read_text())

    rendered = render_recipe_with_context(recipe_yaml)
    into_yaml = _dump_yaml_to_string(rendered)

    assert into_yaml == snapshot

    jolt_physics = test_data / "jolt-physics" / "recipe.yaml"
    variants = (test_data / "jolt-physics" / "ci_support").glob("*.yaml")

    recipe_yaml = load_yaml(jolt_physics.read_text())
    variants = [load_yaml(variant.read_text()) for variant in variants]

    rendered = []
    for v in variants:
        vx = {
            el: v[el][0] for el in v
        }

        rendered.append(render_recipe_with_context(recipe_yaml, vx))

    into_yaml = _dump_yaml_to_string(rendered)

    assert into_yaml == snapshot


def test_multi_source_render(snapshot) -> None:
    jolt_physics = test_data / "jolt-physics" / "sources.yaml"
    variants = (test_data / "jolt-physics" / "ci_support").glob("*.yaml")

    recipe_yaml = load_yaml(jolt_physics.read_text())
    variants = [load_yaml(variant.read_text()) for variant in variants]

    rendered = []
    for v in variants:
        vx = {
            el: v[el][0] for el in v
        }

        print(render_recipe_with_context(recipe_yaml, vx))
