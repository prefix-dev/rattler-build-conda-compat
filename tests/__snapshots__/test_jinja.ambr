# serializer version: 1
# name: test_context_rendering
  '''
  context:
    name: "foo"
    version: "bla"
    name_version: foo-bla
  
  package:
    name: foo
    version: bla
  
  build:
    string: ${{ blas_variant }}${{ hash }}_foo-bla
  
  '''
# ---
# name: test_context_rendering.1
  '''
    - context:
        version: 5.1.0
  
      package:
        name: jolt-physics
        version: 5.1.0
  
      source:
        url: https://github.com/jrouwe/JoltPhysics/archive/refs/tags/v5.1.0.zip
        sha256: 10fcc863ae2b9d48c2f22d8b0204034820e57a55f858b7c388ac9579d8cf4095
  
      requirements:
        build:
          - cxx_compiler_stub
          - c_stdlib_stub
          - cmake
          - ninja
  
      build:
        number: 0
        script:
          - if: win
            then: |
              cmake -GNinja ^
                  %CMAKE_ARGS% ^
                  -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
                  -DBUILD_SHARED_LIBS=ON ^
                  -DCMAKE_BUILD_TYPE=Distribution ^
                  -DCROSS_PLATFORM_DETERMINISTIC=ON ^
                  -DTARGET_VIEWER=OFF ^
                  -DTARGET_SAMPLES=OFF ^
                  -DTARGET_HELLO_WORLD=OFF ^
                  -DTARGET_UNIT_TESTS=OFF ^
                  -DTARGET_PERFORMANCE_TEST=OFF ^
                  -S %SRC_DIR%\Build
              cmake --build . --target install
            else: |
              cmake -GNinja \
                  $CMAKE_ARGS \
                  -DCMAKE_INSTALL_PREFIX=$PREFIX \
                  -DBUILD_SHARED_LIBS=ON \
                  -DCMAKE_BUILD_TYPE=Distribution \
                  -DCROSS_PLATFORM_DETERMINISTIC=ON \
                  -DTARGET_VIEWER=OFF \
                  -DTARGET_SAMPLES=OFF \
                  -DTARGET_HELLO_WORLD=OFF \
                  -DTARGET_UNIT_TESTS=OFF \
                  -DTARGET_PERFORMANCE_TEST=OFF \
                  -S $SRC_DIR/Build
              cmake --build . --target install
  
      about:
        homepage: https://github.com/jrouwe/JoltPhysics
        license: MIT
        license_file: LICENSE
        summary: A multi core friendly rigid body physics and collision detection library.
        description: A multi core friendly rigid body physics and collision detection library. Written in C++. Suitable for games and VR applications. Used by Horizon Forbidden West.
        documentation: https://jrouwe.github.io/JoltPhysics/
        repository: https://github.com/jrouwe/JoltPhysics
  
      extra:
        recipe-maintainers:
          - baszalmstra
    - context:
        version: 5.1.0
  
      package:
        name: jolt-physics
        version: 5.1.0
  
      source:
        url: https://github.com/jrouwe/JoltPhysics/archive/refs/tags/v5.1.0.zip
        sha256: 10fcc863ae2b9d48c2f22d8b0204034820e57a55f858b7c388ac9579d8cf4095
  
      requirements:
        build:
          - cxx_compiler_stub
          - c_stdlib_stub
          - cmake
          - ninja
  
      build:
        number: 0
        script:
          - if: win
            then: |
              cmake -GNinja ^
                  %CMAKE_ARGS% ^
                  -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
                  -DBUILD_SHARED_LIBS=ON ^
                  -DCMAKE_BUILD_TYPE=Distribution ^
                  -DCROSS_PLATFORM_DETERMINISTIC=ON ^
                  -DTARGET_VIEWER=OFF ^
                  -DTARGET_SAMPLES=OFF ^
                  -DTARGET_HELLO_WORLD=OFF ^
                  -DTARGET_UNIT_TESTS=OFF ^
                  -DTARGET_PERFORMANCE_TEST=OFF ^
                  -S %SRC_DIR%\Build
              cmake --build . --target install
            else: |
              cmake -GNinja \
                  $CMAKE_ARGS \
                  -DCMAKE_INSTALL_PREFIX=$PREFIX \
                  -DBUILD_SHARED_LIBS=ON \
                  -DCMAKE_BUILD_TYPE=Distribution \
                  -DCROSS_PLATFORM_DETERMINISTIC=ON \
                  -DTARGET_VIEWER=OFF \
                  -DTARGET_SAMPLES=OFF \
                  -DTARGET_HELLO_WORLD=OFF \
                  -DTARGET_UNIT_TESTS=OFF \
                  -DTARGET_PERFORMANCE_TEST=OFF \
                  -S $SRC_DIR/Build
              cmake --build . --target install
  
      about:
        homepage: https://github.com/jrouwe/JoltPhysics
        license: MIT
        license_file: LICENSE
        summary: A multi core friendly rigid body physics and collision detection library.
        description: A multi core friendly rigid body physics and collision detection library. Written in C++. Suitable for games and VR applications. Used by Horizon Forbidden West.
        documentation: https://jrouwe.github.io/JoltPhysics/
        repository: https://github.com/jrouwe/JoltPhysics
  
      extra:
        recipe-maintainers:
          - baszalmstra
    - context:
        version: 5.1.0
  
      package:
        name: jolt-physics
        version: 5.1.0
  
      source:
        url: https://github.com/jrouwe/JoltPhysics/archive/refs/tags/v5.1.0.zip
        sha256: 10fcc863ae2b9d48c2f22d8b0204034820e57a55f858b7c388ac9579d8cf4095
  
      requirements:
        build:
          - cxx_compiler_stub
          - c_stdlib_stub
          - cmake
          - ninja
  
      build:
        number: 0
        script:
          - if: win
            then: |
              cmake -GNinja ^
                  %CMAKE_ARGS% ^
                  -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
                  -DBUILD_SHARED_LIBS=ON ^
                  -DCMAKE_BUILD_TYPE=Distribution ^
                  -DCROSS_PLATFORM_DETERMINISTIC=ON ^
                  -DTARGET_VIEWER=OFF ^
                  -DTARGET_SAMPLES=OFF ^
                  -DTARGET_HELLO_WORLD=OFF ^
                  -DTARGET_UNIT_TESTS=OFF ^
                  -DTARGET_PERFORMANCE_TEST=OFF ^
                  -S %SRC_DIR%\Build
              cmake --build . --target install
            else: |
              cmake -GNinja \
                  $CMAKE_ARGS \
                  -DCMAKE_INSTALL_PREFIX=$PREFIX \
                  -DBUILD_SHARED_LIBS=ON \
                  -DCMAKE_BUILD_TYPE=Distribution \
                  -DCROSS_PLATFORM_DETERMINISTIC=ON \
                  -DTARGET_VIEWER=OFF \
                  -DTARGET_SAMPLES=OFF \
                  -DTARGET_HELLO_WORLD=OFF \
                  -DTARGET_UNIT_TESTS=OFF \
                  -DTARGET_PERFORMANCE_TEST=OFF \
                  -S $SRC_DIR/Build
              cmake --build . --target install
  
      about:
        homepage: https://github.com/jrouwe/JoltPhysics
        license: MIT
        license_file: LICENSE
        summary: A multi core friendly rigid body physics and collision detection library.
        description: A multi core friendly rigid body physics and collision detection library. Written in C++. Suitable for games and VR applications. Used by Horizon Forbidden West.
        documentation: https://jrouwe.github.io/JoltPhysics/
        repository: https://github.com/jrouwe/JoltPhysics
  
      extra:
        recipe-maintainers:
          - baszalmstra
  
  '''
# ---
# name: test_render_recipe_with_context
  '''
  # yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
  
  context:
    name: mamba
    libmamba_version: "1.5.8"
    libmambapy_version: "1.5.8"
    mamba_version: "1.5.8"
    release: "2024.03.25"
    build_number: 2
  
  recipe:
    name: mamba-split
    version: 1.5.8
  
  source:
    url: https://github.com/mamba-org/mamba/archive/refs/tags/2024.03.25.tar.gz
    sha256: 6ddaf4b0758eb7ca1250f427bc40c2c3ede43257a60bac54e4320a4de66759a6
  
  build:
    number: 2
  
  outputs:
    - package:
        name: libmamba
        version: 1.5.8
      build:
        script:
          - build_mamba.sh
          - 
      requirements:
        build:
          - cxx_compiler_stub
          - cmake
          - ninja
          - 
        host:
          - libsolv >=0.7.23
          - libcurl >=8.4.0
          - fmt
          - 
        run:
          - libsolv >=0.7.23
        run_exports:
          - subpackage_pin libmamba
        ignore_run_exports:
          by_name:
            - spdlog
            - python
      tests:
        - script:
            - if: unix
              then:
                - test -d ${PREFIX}/include/mamba # [unix]
              else:
                - if not exist %LIBRARY_PREFIX%\include\mamba\version.hpp (exit 1) # [win]
  
    - package:
        name: libmambapy
        version: 1.5.8
      build:
        script:
          - build_mamba.sh
          - 
        # string: py_suppython_version_to_build_stringh${{ hash }}_2
      requirements:
        build:
          - cxx_compiler_stub
          - cmake
          - ninja
          - if: build_platform != target_platform
            then:
              - python
              - cross-python_linux-64
              - pybind11
              - pybind11-abi
        host:
          - python
          - nlohmann_json
          - subpackage_pin libmamba
        run:
          - python
          - subpackage_pin libmamba
        run_exports:
          - subpackage_pin libmambapy
        ignore_run_exports:
          by_name:
            - spdlog
      tests:
        - python:
            imports:
              - libmambapy
              - libmambapy.bindings
        - script:
            - python -c "import libmambapy._version; assert libmambapy._version.__version__ == '1.5.8'"
  
    - package:
        name: mamba
        version: 1.5.8
      build:
        script:
          - build_mamba.sh
          - 
          - 
        string: pypython_version_to_build_stringh${{ hash }}_2
        python:
          entry_points:
            - mamba = mamba.mamba:main
      requirements:
        build:
          - if: build_platform != target_platform
            then:
              - python
              - cross-python_linux-64
        run:
          - python
          - conda >=23.9,<24
          - subpackage_pin libmambapy
  
      tests:
        - python:
            imports:
              - mamba
        - script:
            - mamba --help
          # for some reason tqdm doesn't have a proper colorama dependency so pip check fails
          # but that's completely unrelated to mamba
            - python -c "import mamba._version; assert mamba._version.__version__ == '1.5.8'"
  
            - if: linux
              then:
                - test -f ${PREFIX}/etc/profile.d/mamba.sh
              # these tests work when run on win, but for some reason not during conda build
                - mamba create -n test_py2 python=2.7 --dry-run
                - mamba install xtensor xsimd -c conda-forge --dry-run
            - if: unix
              then:
                - test -f ${PREFIX}/condabin/mamba
  
  about:
    homepage: https://github.com/mamba-org/mamba
    license: BSD-3-Clause
    license_file: LICENSE
    license_family: BSD
    summary: A fast drop-in alternative to conda, using libsolv for dependency resolution
    description: |
      # Mamba, the Fast Cross-Platform Package Manager
      env_"MY_ENV_VAR" 
      env_"MY_ENV_VAR" 
      env_exists_"MY_ENV_VAR" 
  
    repository: https://github.com/mamba-org/mamba
  
  '''
# ---
